### PR TITLE
Person Lookup Cache

### PR DESCRIPTION
**Pull request by fdurand**
# Description
Cache the result of person lookup for 1h, it will reduce a lot the ldap search.
Optimize call to person_add in order to reduce the stress on the db.

# Impacts
pfqueue

# Delete branch after merge
YES

# NEWS file entries
## New Features
* Cache person_lookup to reduce the call to ldap_search

### CODE REVIEW

File: conf/chi.conf.defaults [hunk 1]
```diff
@@ -78,3 +78,10 @@ reconnect=60
 #
 # Amount of time from the current time to expire an entry
 expires_in=1d
+
+[storage person_lookup]
+#
+# storage person_lookup.expires_in
+#
+# Amount of time from the current time to expire an entry
+expires_in=1h
```
Comment (julsemaan) [inline conf/chi.conf.defaults]: I'd even go with 24 hours and maybe even more, its not like your lastname will change every hour.

Unless you're called Zammit/Zayme/Zaymitte/ZiCEO

**Comment by extrafu** [comment]
@jrouzierinverse Review please.

File: lib/pf/lookup/person.pm [hunk 2]
```diff
@@ -36,26 +39,31 @@ sub lookup_person {
...
 
     unless (person_exist($pid)) {
-        return "Person $pid is not a registered user!\n";
+        return undef;
     }
-    my ($person, $msg) = $source->search_attributes($pid);
-    if (!$person) {
-       $logger->info("Cannot search attributes for user '$pid'");
-       return "Cannot search attributes for user '$pid'!\n";
+    my $person = $CHI_CACHE->get("$source_id.$pid");
```
Comment (jrouzierinverse) [inline lib/pf/lookup/person.pm:54]: use compute

File: lib/pf/lookup/person.pm [hunk 3]
```diff
@@ -36,26 +39,31 @@ sub lookup_person {
...
     my $source = pf::authentication::getAuthenticationSource($source_id);
     if (!$source) {
        $logger->info("Unable to locate the source $source_id");
-       return "Unable to locate the source $source_id!\n";
+       return undef;
     }
 
     unless (person_exist($pid)) {
-        return "Person $pid is not a registered user!\n";
+        return undef;
```
Comment (jrouzierinverse) [inline lib/pf/lookup/person.pm]: Log the previous message

File: lib/pf/lookup/person.pm [hunk 4]
```diff
@@ -36,26 +39,31 @@ sub lookup_person {
...
+            $logger->info("Successfully did a person lookup for $pid");
+            person_modify($pid, %$person);
+            return undef;
+        }
     }
-
-    $logger->info("Successfully did a person lookup for $pid");
-
-    person_modify($pid, %$person);
+    $logger->info("Already did a person lookup for $pid");
```
Comment (jrouzierinverse) [inline lib/pf/lookup/person.pm:67]: return a true value

File: lib/pf/lookup/person.pm (hunk 2 above)
Comment (fdurand) [reply lib/pf/lookup/person.pm:54]: If i use compute then each time the person has been found in the cache then i will do a person_modify and i don't want to do it.

File: lib/pf/lookup/person.pm (hunk 3 above)
Comment (jrouzierinverse) [reply lib/pf/lookup/person.pm]: Why are you returning a value when the value is not used?

File: lib/pf/lookup/person.pm [hunk 5]
```diff
@@ -36,26 +39,33 @@ sub lookup_person {
     my $logger = get_logger();
     unless (defined $source_id) {
         $logger->info("undefined source id provided");
-        return "Unknown source";
+        return undef;
```
Comment (jrouzierinverse) [inline lib/pf/lookup/person.pm]: missed this one

**Pull request merged**